torch/csrc/jit/codegen/cuda/lower_expr_sort.cpp

```diff
@@ -1326,31 +1412,34 @@ void ExprSegmentationSorter::sort() {
         resetTraversal();
         resetLevels();
 
-        for (auto& group : groups_) {
-          if (group->payload()->merged) {
-            continue;
-          }
-          auto candidates = group->getMergeCandidates(fallback_mode_enabled_);
-          if (candidates.empty()) {
-            continue;
-          }
-
-          auto candidate_it = candidates.begin();
-          while (candidate_it != candidates.end() &&
-                 !supportedMerge(group.get(), *candidate_it)) {
-            candidate_it++;
-          }
-          if (candidate_it == candidates.end()) {
-            continue;
+        if (!horizontallyMergeScalarGroups()) {
+          for (bool preferred_merge_only : {true, false}) {
+            for (auto& group : groups_) {
+              if (group->payload()->merged) {
+                continue;
+              }
+              auto candidates = group->getMergeCandidates(
+                  preferred_merge_only, fallback_mode_enabled_);
+              if (candidates.empty()) {
+                continue;
+              }
+
+              auto candidate_it = candidates.begin();
+              while (candidate_it != candidates.end() &&
+                     !supportedMerge(group.get(), *candidate_it)) {
+                candidate_it++;
+              }
+              if (candidate_it == candidates.end()) {
+                continue;
+              }
+
+              setToMerge(group.get(), *candidate_it);
+            }
+
+            if (to_merge_.size() > 0) {
```

Comment:
nit: The control flow is not easy to follow here as there are 4 levels of loops. Please leave some comment about which loop this `break` exits from and also why.

Reply:
Good point! I actually put the `}` in the wrong place and generated lots of failures during my first try, and it took a while to debug.

Reply:
added